{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" \n",
    "    RNA Alignment Assignment\n",
    "    \n",
    "    Implement each of the functions below using the algorithms covered in class.\n",
    "    You can construct additional functions and data structures but you should not\n",
    "    change the functions' APIs.\n",
    "\n",
    "    You will be graded on the helper function implementations as well as the RNA alignment, although\n",
    "    you do not have to use your helper function.\n",
    "    \n",
    "    *** Make sure to comment out any print statement so as not to interfere with the grading script\n",
    "\"\"\"\n",
    "\n",
    "import sys # DO NOT EDIT THIS\n",
    "from shared import *\n",
    "\n",
    "import time\n",
    "\n",
    "ALPHABET = [TERMINATOR] + BASES\n",
    "\n",
    "def cmp_to_key(mycmp):\n",
    "    'Convert a cmp= function into a key= function'\n",
    "    class K(object):\n",
    "        def __init__(self, obj, *args):\n",
    "            self.obj = obj\n",
    "        def __lt__(self, other):\n",
    "            return mycmp(self.obj, other.obj) < 0\n",
    "        def __gt__(self, other):\n",
    "            return mycmp(self.obj, other.obj) > 0\n",
    "        def __eq__(self, other):\n",
    "            return mycmp(self.obj, other.obj) == 0\n",
    "        def __le__(self, other):\n",
    "            return mycmp(self.obj, other.obj) <= 0  \n",
    "        def __ge__(self, other):\n",
    "            return mycmp(self.obj, other.obj) >= 0\n",
    "        def __ne__(self, other):\n",
    "            return mycmp(self.obj, other.obj) != 0\n",
    "    return K\n",
    "def get_suffix_array(s):\n",
    "    \"\"\"\n",
    "    Naive implementation of suffix array generation (0-indexed). You do not have to implement the\n",
    "    KS Algorithm. Make this code fast enough so you have enough time in Aligner.__init__ (see bottom).\n",
    "\n",
    "    Input:\n",
    "        s: a string of the alphabet ['A', 'C', 'G', 'T'] already terminated by a unique delimiter '$'\n",
    "    \n",
    "    Output: list of indices representing the suffix array\n",
    "\n",
    "    >>> get_suffix_array('GATAGACA$')\n",
    "    [8, 7, 5, 3, 1, 6, 4, 0, 2]\n",
    "    \"\"\"\n",
    "    def compare(i, j):\n",
    "        c = 0\n",
    "        while s[i + c] == s[j + c]:\n",
    "            c += 1\n",
    "        if s[i + c] > s[j + c]:\n",
    "            return 1\n",
    "        elif s[i + c] < s[j + c]:\n",
    "            return -1\n",
    "        else:\n",
    "            return 0\n",
    "    \n",
    "    buckets = {}\n",
    "    for i in range(len(s)):\n",
    "        if s[i:i+100] in buckets:\n",
    "            buckets[s[i:i+100]].append(i)\n",
    "        else:\n",
    "            buckets[s[i:i+100]] = [i]\n",
    "\n",
    "    keys = list(buckets.keys())\n",
    "    for i in range(len(keys)):\n",
    "        item = keys[i]\n",
    "        if len(buckets[item]) > 1:\n",
    "            buckets[item].sort(key=cmp_to_key(compare))\n",
    "\n",
    "    result = []\n",
    "    for item in sorted(keys):\n",
    "        result.extend(buckets[item])\n",
    "\n",
    "    return result\n",
    "\n",
    "def get_bwt(s, sa):\n",
    "    \"\"\"\n",
    "    Input:\n",
    "        s: a string terminated by a unique delimiter '$'\n",
    "        sa: the suffix array of s\n",
    "\n",
    "    Output:\n",
    "        L: BWT of s as a string\n",
    "    \"\"\"\n",
    "    return ''.join([s[i-1] for i in sa])\n",
    "\n",
    "def get_F(L):\n",
    "    \"\"\"\n",
    "    Input: L = get_bwt(s)\n",
    "\n",
    "    Output: F, first column in Pi_sorted\n",
    "    \"\"\"\n",
    "    return ''.join(sorted(list(L)))\n",
    "\n",
    "def get_M(F):\n",
    "    \"\"\"\n",
    "    Returns the helper data structure M (using the notation from class). M is a dictionary that maps character\n",
    "    strings to start indices. i.e. M[c] is the first occurrence of \"c\" in F.\n",
    "\n",
    "    If a character \"c\" does not exist in F, you may set M[c] = -1\n",
    "    \"\"\"\n",
    "    M = {}\n",
    "    for i in range(len(F)):\n",
    "        char = F[i]\n",
    "        if char not in M:\n",
    "            M[char] = i\n",
    "    for char in ALPHABET:\n",
    "        if char not in M:\n",
    "            M[char] = -1\n",
    "    return M\n",
    "\n",
    "def get_occ(L):\n",
    "    \"\"\"\n",
    "    Returns the helper data structure OCC (using the notation from class). OCC should be a dictionary that maps \n",
    "    string character to a list of integers. If c is a string character and i is an integer, then OCC[c][i] gives\n",
    "    the number of occurrences of character \"c\" in the bwt string up to and including index i\n",
    "    \"\"\"\n",
    "    occ = {i: [0] for i in ALPHABET}\n",
    "    for i in L:\n",
    "        for j in occ:\n",
    "            if i != j:\n",
    "                occ[j].append(occ[j][-1])\n",
    "            else:\n",
    "                occ[i].append(occ[i][-1]+1)\n",
    "    for i in occ:\n",
    "        occ[i].pop(0)\n",
    "    return occ\n",
    "\n",
    "def construct_L(M, occ):\n",
    "    length = 0\n",
    "    for char in occ:\n",
    "        length += occ[char][-1]\n",
    "    L = ['']*length\n",
    "    for char in occ:\n",
    "        for i in range(occ[char][-1]):\n",
    "            L[occ[char].index(i+1)] = char\n",
    "    return ''.join(L)\n",
    "\n",
    "def exact_suffix_matches(p, M, occ):\n",
    "    \"\"\"\n",
    "    Find the positions within the suffix array sa of the longest possible suffix of p \n",
    "    that is a substring of s (the original string).\n",
    "    \n",
    "    Note that such positions must be consecutive, so we want the range of positions.\n",
    "\n",
    "    Input:\n",
    "        p: the pattern string\n",
    "        M, occ: buckets and repeats information used by sp, ep\n",
    "\n",
    "    Output: a tuple (range, length)\n",
    "        range: a tuple (start inclusive, end exclusive) of the indices in sa that contains\n",
    "            the longest suffix of p as a prefix. range=None if no indices matches any suffix of p\n",
    "        length: length of the longest suffix of p found in s. length=0 if no indices matches any suffix of p\n",
    "\n",
    "        An example return value would be ((2, 5), 7). This means that p[len(p) - 7 : len(p)] is\n",
    "        found in s and matches positions 2, 3, and 4 in the suffix array.\n",
    "\n",
    "    >>> s = 'ACGT' * 10 + '$'\n",
    "    >>> sa = get_suffix_array(s)\n",
    "    >>> sa\n",
    "    [40, 36, 32, 28, 24, 20, 16, 12, 8, 4, 0, 37, 33, 29, 25, 21, 17, 13, 9, 5, 1, 38, 34, 30, 26, 22, 18, 14, 10, 6, 2, 39, 35, 31, 27, 23, 19, 15, 11, 7, 3]\n",
    "    >>> L = get_bwt(s, sa)\n",
    "    >>> L\n",
    "    'TTTTTTTTTT$AAAAAAAAAACCCCCCCCCCGGGGGGGGGG'\n",
    "    >>> F = get_F(L)\n",
    "    >>> F\n",
    "    '$AAAAAAAAAACCCCCCCCCCGGGGGGGGGGTTTTTTTTTT'\n",
    "    >>> M = get_M(F)\n",
    "    >>> sorted(M.items())\n",
    "    [('$', 0), ('A', 1), ('C', 11), ('G', 21), ('T', 31)]\n",
    "    >>> occ = get_occ(L)\n",
    "    >>> type(occ) == dict, type(occ['$']) == list, type(occ['$'][0]) == int\n",
    "    (True, True, True)\n",
    "    >>> occ['$']\n",
    "    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]\n",
    "    >>> exact_suffix_matches('ACTGA', M, occ)\n",
    "    ((1, 11), 1)\n",
    "    >>> exact_suffix_matches('$', M, occ)\n",
    "    ((0, 1), 1)\n",
    "    >>> exact_suffix_matches('AA', M, occ)\n",
    "    ((1, 11), 1)\n",
    "    \"\"\"\n",
    "    \n",
    "    # initialize sp, ep\n",
    "    length = len(p)\n",
    "    last_char = p[-1]\n",
    "    sp = M[last_char]\n",
    "    if sp == -1:\n",
    "        return (None, 0)\n",
    "    \n",
    "    # find next(last_char)\n",
    "    nxt = float('inf')\n",
    "    nxt_item = None\n",
    "    for item in M:\n",
    "        if nxt > M[item] > M[last_char] and item != last_char:\n",
    "            nxt = M[item]\n",
    "            nxt_item = item\n",
    "    if nxt_item == None:\n",
    "        ep = len(occ['$']) - 1\n",
    "    else:\n",
    "        ep = nxt - 1\n",
    "    \n",
    "    # changed for loop a bit, only works on strings len >= 2\n",
    "    # for strings of len 1, it skips to the end and works\n",
    "    for i in range(length-2,-1,-1):\n",
    "        sp_ph = M[p[i]] + occ[p[i]][sp-1]\n",
    "        ep_ph = M[p[i]] + occ[p[i]][ep]-1\n",
    "        if sp_ph > ep_ph:\n",
    "            return ((sp,ep+1),length-i-1)\n",
    "        sp = sp_ph\n",
    "        ep = ep_ph\n",
    "    return ((sp,ep+1), length)\n",
    "\n",
    "def bowtie(p, M, occ):\n",
    "    num_mismatches = 0\n",
    "    length = len(p)\n",
    "    last_char = p[-1]\n",
    "    sp = M[last_char]\n",
    "    if sp == -1:\n",
    "        return (None, 0)\n",
    "    \n",
    "    # find next(last_char)\n",
    "    nxt = float('inf')\n",
    "    nxt_item = None\n",
    "    for item in M:\n",
    "        if nxt > M[item] > M[last_char] and item != last_char:\n",
    "            nxt = M[item]\n",
    "            nxt_item = item\n",
    "    if nxt_item == None:\n",
    "        ep = len(occ['$']) - 1\n",
    "    else:\n",
    "        ep = nxt - 1\n",
    "    \n",
    "    # changed for loop a bit, only works on strings len >= 2\n",
    "    # for strings of len 1, it skips to the end and works\n",
    "    for i in range(length-2,-1,-1):\n",
    "        sp_ph = M[p[i]] + occ[p[i]][sp-1]\n",
    "        ep_ph = M[p[i]] + occ[p[i]][ep]-1\n",
    "\n",
    "        poss_switches = ALPHABET.copy()\n",
    "        poss_switches.remove(p[i])\n",
    "        is_mismatch = False\n",
    "        while sp_ph > ep_ph and poss_switches:\n",
    "            new_char = poss_switches.pop()\n",
    "            sp_ph = M[new_char] + occ[new_char][sp-1]\n",
    "            ep_ph = M[new_char] + occ[new_char][ep]-1\n",
    "            is_mismatch = True\n",
    "\n",
    "        num_mismatches += is_mismatch\n",
    "        sp = sp_ph\n",
    "        ep = ep_ph\n",
    "\n",
    "    return ((sp,ep+1), num_mismatches)\n",
    "\n",
    "def suffix_replacer(hits):\n",
    "    (start, end), length = hits\n",
    "    genome_seqs = []\n",
    "    for i in range(start,end):\n",
    "        index = a.genome_sa[i]\n",
    "        genome_seqs.append((index,index+length))\n",
    "    return genome_seqs\n",
    "\n",
    "MIN_INTRON_SIZE = 20\n",
    "MAX_INTRON_SIZE = 10000\n",
    "\n",
    "class Aligner: \n",
    "    def __init__(self, genome_sequence, known_genes):\n",
    "        \"\"\"\n",
    "        Initializes the aligner. Do all time intensive set up here. i.e. build suffix array.\n",
    "\n",
    "        genome_sequence: a string (NOT TERMINATED BY '$') representing the bases of the of the genome\n",
    "        known_genes: a python set of Gene objects (see shared.py) that represent known genes. You can get the isoforms \n",
    "                     and exons from a Gene object\n",
    "\n",
    "        Time limit: 500 seconds maximum on the provided data. Note that our server is probably faster than your machine, \n",
    "                    so don't stress if you are close. Server is 1.25 times faster than the i7 CPU on my computer\n",
    "\n",
    "        \"\"\"\n",
    "        with open('genome.fa') as f:\n",
    "            genome_sequence = f.read().split()[1].strip() + '$'\n",
    "        last_time = time.time()\n",
    "        self.genome_sa = get_suffix_array(genome_sequence)\n",
    "        self.genome_L = get_bwt(genome_sequence, self.genome_sa)\n",
    "        self.genome_M = get_M(get_F(self.genome_L))\n",
    "        self.genome_occ = get_occ(self.genome_L)\n",
    "        \n",
    "        print('FM index of genome: ' + str(time.time() - last_time))\n",
    "        last_time = time.time()\n",
    "        \n",
    "        isos_sa = {}\n",
    "        isos_L = {}\n",
    "        isos_M = {}\n",
    "        isos_occ = {}\n",
    "        isos = {}\n",
    "        iso_exon_starts = {}\n",
    "        iso_names = []\n",
    "        known_genes = []\n",
    "        with open('genes.tab') as f:\n",
    "            curr = f.readline().split()\n",
    "            gene_id = curr[1]\n",
    "            isoform_id = ''\n",
    "            isoforms = []\n",
    "            exons = []\n",
    "            first_gene = -1\n",
    "            while curr:\n",
    "                if curr[0] == 'gene':\n",
    "                    if exons:\n",
    "                        isoforms.append(Isoform(isoform_id,exons))\n",
    "                    if first_gene != -1:\n",
    "                        known_genes.append(Gene(gene_id, isoforms))\n",
    "                    else:\n",
    "                        first_gene = 0\n",
    "                    gene_id = curr[1]\n",
    "                    isoforms = []\n",
    "                    first_iso = -1\n",
    "                elif curr[0] == 'isoform':\n",
    "                    if first_iso != -1:\n",
    "                        isoforms.append(Isoform(isoform_id, exons))\n",
    "                    else:\n",
    "                        first_iso = 0\n",
    "                    isoform_id = curr[1]\n",
    "                    exons = []\n",
    "                elif curr[0] == 'exon':\n",
    "                    exons.append(Exon(curr[1],int(curr[2]),int(curr[3])))\n",
    "                curr = f.readline().split()\n",
    "        \n",
    "        for gene in known_genes:\n",
    "            for isoform in gene.isoforms:\n",
    "                iso_names.append(isoform.id)\n",
    "                spliced_isoform = ''\n",
    "                i = isoform.id\n",
    "                isos[i] = isoform\n",
    "                prev = 0\n",
    "                iso_exon_starts[i] = []\n",
    "                curr = 0\n",
    "                for exon in isoform.exons:\n",
    "                    exon_str = genome_sequence[exon.start:exon.end]\n",
    "                    spliced_isoform += exon_str\n",
    "                    curr += len(exon_str)\n",
    "                    iso_exon_starts[i].append(curr)\n",
    "                spliced_isoform += '$'\n",
    "                isos_sa[i] = get_suffix_array(spliced_isoform)\n",
    "                isos_L[i] = get_bwt(spliced_isoform, isos_sa[i])\n",
    "                isos_M[i] = get_M(get_F(isos_L[i]))\n",
    "                isos_occ[i] = get_occ(isos_L[i])\n",
    "                print('FM index of isoform, length %d: ' % len(spliced_isoform) + str(time.time() - last_time))\n",
    "                last_time = time.time()\n",
    "        \n",
    "        self.isos_sa = isos_sa\n",
    "        self.isos_L = isos_L\n",
    "        self.isos_M = isos_M\n",
    "        self.isos_occ = isos_occ\n",
    "        self.iso_names = iso_names\n",
    "        self.isos = isos\n",
    "        self.iso_exon_starts = iso_exon_starts\n",
    "\n",
    "    def align(self, read_sequence):\n",
    "        \"\"\"\n",
    "        Returns an alignment to the genome sequence. An alignment is a list of pieces. \n",
    "        Each piece consists of a start index in the read, a start index in the genome, and a length \n",
    "        indicating how many bases are aligned in this piece. Note that mismatches are count as \"aligned\".\n",
    "\n",
    "        Note that <read_start_2> >= <read_start_1> + <length_1>. If your algorithm produces an alignment that \n",
    "        violates this, we will remove pieces from your alignment arbitrarily until consecutive pieces \n",
    "        satisfy <read_start_2> >= <read_start_1> + <length_1>\n",
    "\n",
    "        Return value must be in the form (also see the project pdf):\n",
    "        [(<read_start_1>, <reference_start_1, length_1), (<read_start_2>, <reference_start_2, length_2), ...]\n",
    "\n",
    "        If no good matches are found: return the best match you can find or return []\n",
    "\n",
    "        Time limit: 0.5 seconds per read on average on the provided data.\n",
    "        \"\"\"\n",
    "        \n",
    "        # bowtie, read to isoforms\n",
    "        # first see if the read matches a known isoform for under 6 mismatches\n",
    "        min_mismatches = float('inf')\n",
    "        best_match = None\n",
    "        for iso_name in self.iso_names:\n",
    "            match_range, num_mismatches = bowtie(read_sequence, self.isos_M[iso_name], self.isos_occ[iso_name])\n",
    "            if num_mismatches < min_mismatches:\n",
    "                min_mismatches = num_mismatches\n",
    "                best_match = (iso_name, match_range)\n",
    "        \n",
    "        alignment = []\n",
    "        seqlength = len(read_sequence)\n",
    "        \n",
    "        #Transcriptome Alignment\n",
    "        if min_mismatches <= 6 and best_match:\n",
    "            # we have a match, direct to transcriptome\n",
    "            # if we have multiple matches, it'll be weird but just take the first one\n",
    "            iso_name, match_range = best_match\n",
    "            curr = self.isos_sa[iso_name][match_range[0]] # start pt in spliced isoform\n",
    "            print(curr)\n",
    "            exon_starts = self.iso_exon_starts[iso_name]\n",
    "            exon_num = 0\n",
    "            curr_pt = 0\n",
    "            for i in range(len(exon_starts)):\n",
    "                if curr < exon_starts[i]:\n",
    "                    exon_num = i\n",
    "                    break\n",
    "            exons = self.isos[iso_name].exons\n",
    "            print(exons)\n",
    "            print(exon_starts)\n",
    "            read_pt = 0\n",
    "            curr_pt = exons[exon_num].start + curr\n",
    "            while seqlength:\n",
    "                offset = exons[exon_num].end-curr_pt\n",
    "                diff = min(offset, seqlength)\n",
    "                alignment.append((read_pt, curr_pt, diff))\n",
    "                seqlength -= diff\n",
    "                curr_pt += diff\n",
    "                read_pt += diff\n",
    "                exon_num += 1\n",
    "            return alignment\n",
    "        #Genome Alignment\n",
    "        else:\n",
    "            min_mis = 7\n",
    "            best_hits = []\n",
    "            hits = bowtie(read, a.genome_M, a.genome_occ)\n",
    "            curr_mis = hits[1]\n",
    "            if curr_mis < min_mis:\n",
    "                min_mis = curr_mis\n",
    "                best_hits = [(0, suffix_replacer(hits)[0][0], 50)]\n",
    "            for offset in range(10, 41):\n",
    "                seed1, seed2 = read[:offset], read[offset:]\n",
    "                hits1 = bowtie(seed1, a.genome_M, a.genome_occ)\n",
    "                hits2 = bowtie(seed2, a.genome_M, a.genome_occ)\n",
    "                curr_mis = hits1[1] + hits2[1]\n",
    "                if curr_mis < min_mis:\n",
    "                    hits1 = suffix_replacer(hits1)\n",
    "                    hits2 = suffix_replacer(hits2)\n",
    "                    for seq1 in hits1:\n",
    "                        for seq2 in hits2:\n",
    "                            if MIN_INTRON_SIZE < seq2[0] - seq1[1] < MAX_INTRON_SIZE:\n",
    "                                min_mis = curr_mis\n",
    "                                seq1len = len(seq1)\n",
    "                                seq2len = len(seq2)\n",
    "                                best_hits = [(0, seq1[0], seq1len), (seq1len, seq2[0], seq2len)]\n",
    "                                break\n",
    "            for offset1 in range(10,31):\n",
    "                for offset2 in range(10+offset1, 41):\n",
    "                    seed1, seed2,seed3 = read[:offset1], read[offset1:offset2], read[offset2:]\n",
    "                    hits1 = bowtie(seed1, a.genome_M, a.genome_occ)\n",
    "                    hits2 = bowtie(seed2, a.genome_M, a.genome_occ)\n",
    "                    hits3 = bowtie(seed3, a.genome_M, a.genome_occ)\n",
    "                    curr_mis = hits1[1] + hits2[1] + hits3[1]\n",
    "                    if curr_mis < min_mis:\n",
    "                        hits1 = suffix_replacer(hits1)\n",
    "                        hits2 = suffix_replacer(hits2)\n",
    "                        hits3 = suffix_replacer(hits3)\n",
    "                        for seq1 in hits1:\n",
    "                            for seq2 in hits2:\n",
    "                                for seq3 in hits3:\n",
    "                                    if MIN_INTRON_SIZE < seq2[0] - seq1[1] < MAX_INTRON_SIZE and MIN_INTRON_SIZE < seq3[0] - seq2[1] < MAX_INTRON_SIZE:\n",
    "                                        min_mis = curr_mis\n",
    "                                        seq1len = len(seq1)\n",
    "                                        seq2len = len(seq2)\n",
    "                                        seq3len = len(seq3)\n",
    "                                        best_hits = [(0, seq1[0], seq1len), (seq1len, seq2[0], seq2len), (seq1len+seq2len, seq3[0], seq3len)]\n",
    "                                        break\n",
    "            return best_hits\n",
    "        return []\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FM index of genome: 280.10188007354736\n",
      "FM index of isoform, length 635: 1.616556167602539\n",
      "FM index of isoform, length 666: 0.0036821365356445312\n",
      "FM index of isoform, length 501: 0.0026731491088867188\n",
      "FM index of isoform, length 1265: 0.006295204162597656\n",
      "FM index of isoform, length 107: 0.0004401206970214844\n",
      "FM index of isoform, length 2191: 0.008221864700317383\n",
      "FM index of isoform, length 490: 0.001669168472290039\n",
      "FM index of isoform, length 107: 0.0003902912139892578\n",
      "FM index of isoform, length 851: 0.002853870391845703\n",
      "FM index of isoform, length 457: 0.0014488697052001953\n",
      "FM index of isoform, length 89: 0.0002911090850830078\n",
      "FM index of isoform, length 84: 0.00028204917907714844\n",
      "FM index of isoform, length 756: 0.0027718544006347656\n",
      "FM index of isoform, length 479: 0.0015957355499267578\n",
      "FM index of isoform, length 507: 0.001531839370727539\n",
      "FM index of isoform, length 2148: 1.8316681385040283\n",
      "FM index of isoform, length 601: 0.0019829273223876953\n",
      "FM index of isoform, length 607: 0.002260923385620117\n",
      "FM index of isoform, length 2604: 0.010123968124389648\n",
      "FM index of isoform, length 3950: 0.0305330753326416\n",
      "FM index of isoform, length 5192: 0.06688404083251953\n",
      "FM index of isoform, length 681: 0.0029249191284179688\n",
      "FM index of isoform, length 844: 0.0038890838623046875\n",
      "FM index of isoform, length 5316: 0.024385929107666016\n",
      "FM index of isoform, length 750: 0.00249481201171875\n",
      "FM index of isoform, length 107: 0.00035881996154785156\n",
      "FM index of isoform, length 4683: 0.0191192626953125\n",
      "FM index of isoform, length 579: 0.0020580291748046875\n",
      "FM index of isoform, length 338: 0.0011599063873291016\n",
      "FM index of isoform, length 8118: 0.04448890686035156\n",
      "FM index of isoform, length 578: 0.002233743667602539\n",
      "FM index of isoform, length 647: 0.0025212764739990234\n",
      "FM index of isoform, length 667: 0.004827976226806641\n",
      "FM index of isoform, length 1988: 0.01195979118347168\n",
      "FM index of isoform, length 1594: 0.008231163024902344\n",
      "FM index of isoform, length 1930: 0.010548830032348633\n",
      "FM index of isoform, length 676: 0.0028078556060791016\n",
      "FM index of isoform, length 549: 0.0018761157989501953\n",
      "FM index of isoform, length 502: 0.0016779899597167969\n",
      "FM index of isoform, length 1181: 0.004683971405029297\n",
      "FM index of isoform, length 498: 0.0019159317016601562\n",
      "FM index of isoform, length 1537: 0.0067348480224609375\n",
      "FM index of isoform, length 447: 0.001558065414428711\n",
      "FM index of isoform, length 86: 0.00033473968505859375\n",
      "FM index of isoform, length 962: 0.0045850276947021484\n",
      "FM index of isoform, length 1101: 0.0055086612701416016\n",
      "FM index of isoform, length 217: 0.0010538101196289062\n",
      "FM index of isoform, length 711: 0.004309892654418945\n",
      "FM index of isoform, length 561: 0.004189968109130859\n",
      "FM index of isoform, length 780: 0.005093097686767578\n",
      "FM index of isoform, length 165: 0.0006189346313476562\n",
      "FM index of isoform, length 564: 0.0022211074829101562\n",
      "FM index of isoform, length 217: 0.0008130073547363281\n",
      "FM index of isoform, length 1330: 0.005029201507568359\n",
      "FM index of isoform, length 778: 0.004292964935302734\n",
      "FM index of isoform, length 1082: 0.008274078369140625\n",
      "FM index of isoform, length 836: 0.005115032196044922\n",
      "FM index of isoform, length 512: 0.002260923385620117\n",
      "FM index of isoform, length 1995: 0.009768247604370117\n",
      "FM index of isoform, length 765: 0.0031249523162841797\n",
      "FM index of isoform, length 284: 0.0009658336639404297\n",
      "FM index of isoform, length 378: 0.0015571117401123047\n",
      "FM index of isoform, length 295: 0.0010418891906738281\n",
      "FM index of isoform, length 3351: 0.015317201614379883\n",
      "FM index of isoform, length 2742: 0.012563943862915039\n",
      "FM index of isoform, length 685: 0.0028638839721679688\n",
      "FM index of isoform, length 607: 0.0033621788024902344\n",
      "FM index of isoform, length 3815: 0.016798019409179688\n",
      "FM index of isoform, length 4076: 0.013679981231689453\n",
      "FM index of isoform, length 292: 0.0009388923645019531\n",
      "FM index of isoform, length 877: 0.004178047180175781\n",
      "FM index of isoform, length 549: 0.002789020538330078\n",
      "FM index of isoform, length 572: 0.0016601085662841797\n",
      "FM index of isoform, length 352: 0.0011429786682128906\n",
      "FM index of isoform, length 187: 0.0006608963012695312\n",
      "FM index of isoform, length 705: 0.009903192520141602\n",
      "FM index of isoform, length 369: 0.0036199092864990234\n",
      "FM index of isoform, length 471: 0.0019876956939697266\n",
      "FM index of isoform, length 715: 0.003119230270385742\n",
      "FM index of isoform, length 524: 0.0014522075653076172\n",
      "FM index of isoform, length 555: 0.0019178390502929688\n",
      "FM index of isoform, length 1583: 0.00503993034362793\n",
      "FM index of isoform, length 518: 0.001497030258178711\n",
      "FM index of isoform, length 695: 0.0021979808807373047\n",
      "FM index of isoform, length 642: 0.0017349720001220703\n",
      "FM index of isoform, length 561: 0.0014421939849853516\n",
      "FM index of isoform, length 619: 0.0017557144165039062\n",
      "FM index of isoform, length 573: 0.001344919204711914\n",
      "FM index of isoform, length 555: 0.0012979507446289062\n",
      "FM index of isoform, length 3469: 0.011523008346557617\n",
      "FM index of isoform, length 528: 0.001361846923828125\n",
      "FM index of isoform, length 1850: 0.006757020950317383\n",
      "FM index of isoform, length 109: 0.0002989768981933594\n",
      "FM index of isoform, length 932: 0.0031349658966064453\n",
      "FM index of isoform, length 211: 0.0010418891906738281\n",
      "FM index of isoform, length 2048: 0.010667085647583008\n",
      "FM index of isoform, length 244: 0.0013189315795898438\n",
      "FM index of isoform, length 578: 0.0032720565795898438\n",
      "FM index of isoform, length 445: 0.0018620491027832031\n",
      "FM index of isoform, length 160: 0.0006458759307861328\n",
      "FM index of isoform, length 382: 0.0016438961029052734\n",
      "FM index of isoform, length 421: 0.0015621185302734375\n",
      "FM index of isoform, length 468: 0.0015170574188232422\n",
      "FM index of isoform, length 1353: 0.006707906723022461\n",
      "FM index of isoform, length 662: 0.0024819374084472656\n",
      "FM index of isoform, length 3256: 0.015604019165039062\n",
      "FM index of isoform, length 1600: 0.004832029342651367\n",
      "FM index of isoform, length 101: 0.00029277801513671875\n",
      "FM index of isoform, length 608: 0.0014638900756835938\n",
      "FM index of isoform, length 87: 0.0002219676971435547\n",
      "FM index of isoform, length 268: 0.0008630752563476562\n",
      "FM index of isoform, length 606: 0.002504110336303711\n",
      "FM index of isoform, length 1971: 0.009788990020751953\n",
      "FM index of isoform, length 2032: 0.011630058288574219\n",
      "FM index of isoform, length 2131: 0.011081218719482422\n",
      "FM index of isoform, length 2233: 0.015590906143188477\n",
      "FM index of isoform, length 2415: 0.023441076278686523\n",
      "FM index of isoform, length 1632: 0.008220911026000977\n",
      "FM index of isoform, length 2543: 0.011964797973632812\n",
      "FM index of isoform, length 2981: 0.02383279800415039\n",
      "FM index of isoform, length 1095: 0.004836082458496094\n",
      "FM index of isoform, length 116: 0.0004589557647705078\n",
      "FM index of isoform, length 81: 0.0003662109375\n",
      "FM index of isoform, length 442: 0.0014431476593017578\n",
      "FM index of isoform, length 550: 0.0019888877868652344\n",
      "FM index of isoform, length 2341: 0.012747764587402344\n"
     ]
    }
   ],
   "source": [
    "a = Aligner(None, None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "def outer_align(read_sequence):\n",
    "        \"\"\"\n",
    "        Returns an alignment to the genome sequence. An alignment is a list of pieces. \n",
    "        Each piece consists of a start index in the read, a start index in the genome, and a length \n",
    "        indicating how many bases are aligned in this piece. Note that mismatches are count as \"aligned\".\n",
    "\n",
    "        Note that <read_start_2> >= <read_start_1> + <length_1>. If your algorithm produces an alignment that \n",
    "        violates this, we will remove pieces from your alignment arbitrarily until consecutive pieces \n",
    "        satisfy <read_start_2> >= <read_start_1> + <length_1>\n",
    "\n",
    "        Return value must be in the form (also see the project pdf):\n",
    "        [(<read_start_1>, <reference_start_1, length_1), (<read_start_2>, <reference_start_2, length_2), ...]\n",
    "\n",
    "        If no good matches are found: return the best match you can find or return []\n",
    "\n",
    "        Time limit: 0.5 seconds per read on average on the provided data.\n",
    "        \"\"\"\n",
    "        \n",
    "        # bowtie, read to isoforms\n",
    "        # first see if the read matches a known isoform for under 6 mismatches\n",
    "        min_mismatches = float('inf')\n",
    "        best_match = None\n",
    "        for iso_name in a.iso_names:\n",
    "            match_range, num_mismatches = bowtie(read_sequence, a.isos_M[iso_name], a.isos_occ[iso_name])\n",
    "            if num_mismatches < min_mismatches:\n",
    "                min_mismatches = num_mismatches\n",
    "                best_match = (iso_name, match_range)\n",
    "        \n",
    "        alignment = []\n",
    "        seqlength = len(read_sequence)\n",
    "        \n",
    "        if min_mismatches <= 6 and best_match:\n",
    "            # we have a match, direct to transcriptome\n",
    "            # if we have multiple matches, it'll be weird but just take the first one\n",
    "            iso_name, match_range = best_match\n",
    "            curr = a.isos_sa[iso_name][match_range[0]] # start pt in spliced isoform\n",
    "            print(curr)\n",
    "            exon_starts = a.iso_exon_starts[iso_name]\n",
    "            exon_num = 0\n",
    "            curr_pt = 0\n",
    "            for i in range(len(exon_starts)):\n",
    "                if curr < exon_starts[i]:\n",
    "                    exon_num = i\n",
    "                    break\n",
    "            exons = a.isos[iso_name].exons\n",
    "            print(exons)\n",
    "            print(exon_starts)\n",
    "            read_pt = 0\n",
    "            curr_pt = exons[exon_num].start + curr\n",
    "            while seqlength:\n",
    "                offset = exons[exon_num].end-curr_pt\n",
    "                diff = min(offset, seqlength)\n",
    "                alignment.append((read_pt, curr_pt, diff))\n",
    "                seqlength -= diff\n",
    "                curr_pt += diff\n",
    "                read_pt += diff\n",
    "                exon_num += 1\n",
    "        return alignment\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('genome.fa') as f:\n",
    "    f.readline()\n",
    "    genome = f.readline()[:-1]+'$'\n",
    "reads = []\n",
    "with open('reads.fa') as f:\n",
    "    while f.readline():\n",
    "        reads.append(f.readline()[:-1])        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "known_genes = []\n",
    "unknown_genes = []\n",
    "from shared import *\n",
    "with open('genes.tab') as f:\n",
    "    curr = f.readline().split()\n",
    "    gene_id = curr[1]\n",
    "    isoform_id = ''\n",
    "    isoforms = []\n",
    "    exons = []\n",
    "    first_gene = -1\n",
    "    while curr:\n",
    "        if curr[0] == 'gene':\n",
    "            if exons:\n",
    "                isoforms.append(Isoform(isoform_id,exons))\n",
    "            if first_gene != -1:\n",
    "                known_genes.append(Gene(gene_id, isoforms))\n",
    "            else:\n",
    "                first_gene = 0\n",
    "            gene_id = curr[1]\n",
    "            isoforms = []\n",
    "            first_iso = -1\n",
    "        elif curr[0] == 'isoform':\n",
    "            if first_iso != -1:\n",
    "                isoforms.append(Isoform(isoform_id, exons))\n",
    "            else:\n",
    "                first_iso = 0\n",
    "            isoform_id = curr[1]\n",
    "            exons = []\n",
    "        elif curr[0] == 'exon':\n",
    "            exons.append(Exon(curr[1],int(curr[2]),int(curr[3])))\n",
    "        else:\n",
    "            break\n",
    "        curr = f.readline().split()\n",
    "    isoforms.append(Isoform(isoform_id, exons))\n",
    "    known_genes.append(Gene(gene_id,isoforms))\n",
    "    isoforms = []\n",
    "    first_gene = -1\n",
    "    while curr:\n",
    "        if curr[0] == 'unknown gene':\n",
    "            if exons:\n",
    "                isoforms.append(Isoform(isoform_id,exons))\n",
    "            if first_gene != -1:\n",
    "                genes.append(Gene(gene_id, isoforms))\n",
    "            else:\n",
    "                first_gene = 0\n",
    "            gene_id = curr[1]\n",
    "            isoforms = []\n",
    "            first_iso = -1\n",
    "        elif curr[0] == 'unknown isoform':\n",
    "            if first_iso != -1:\n",
    "                isoforms.append(Isoform(isoform_id, exons))\n",
    "            else:\n",
    "                first_iso = 0\n",
    "            isoform_id = curr[1]\n",
    "            exons = []\n",
    "        elif curr[0] == 'unknown exon':\n",
    "            exons.append(Exon(curr[1],int(curr[2]),int(curr[3])))\n",
    "        curr = f.readline().split()\n",
    "    isoforms.append(Isoform(isoform_id,exons))\n",
    "    unknown_genes.append(Gene(gene_id, isoforms))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "known_isoforms = set()\n",
    "unknown_isoforms = set()\n",
    "for gene in known_genes:\n",
    "    for isoform in gene.isoforms:\n",
    "        known_isoforms.add(isoform)\n",
    "for gene in unknown_genes:\n",
    "    for isoform in gene.isoforms:\n",
    "        unknown_isoforms.add(isoform)\n",
    "from evaluation import *\n",
    "genome_isoform_offsets = index_isoform_locations(known_isoforms, unknown_isoforms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TAAAGGAGTTGTCTAGGGGAGAGCCTTGGATCTGAGAACGCTGTCTGGGCTTGGTAGCAAGAGCTGGTATTTCACAGCAGCGACCGCCTCACAGACGAGCTCTGGGAAGCAGGAGGGAGATTCTCTCTTAGGGTGCCACTACTATGGAGACACAGCTGGGCTGAACA$GATTTGCTTCAAACGACAACAAGAGAGACAGAAAGTATCTGGTTCAACCGCTGCCCGAGCAGGCAGGCACCAGAAGAGCTAGCAGCCAATCCGGCCATGCTCGAAGCCAGCTCTGCAGCTCAGCCAATCAGTGACATCATTGTTTCTTTAGCTTGCCCATGGTGATGTGAAGATGAGAAGAAATAGCAAGGCCCAACCAGTTCTTCATCTGGAGACAGTTCAACGTTCTGCAAACCAGTCTTCAGGAATTTTACTGGGATAATTATCCAAATAAATTGCAAGCATTCTATCCAAATGACATCCATAAAAATTCATCAGAGAGCATTTTACTACTGAGCTGCAAAGGAAAAACTTAAAATGGATATGAAAAGTGAAGAAAGTGATCATAGGAGAAAACCATTTCAGTGTCTGATGATTTGTGAGCCTCAAGTGACTACTCTGTGGAATCATCTGTCCTTGCCTTTGAATTAGCTTTCCCTAGGCCATGAGTCAACATTAGTAGTTTATACATCCAAATTGTGACAAGAGCACCTCAAAGGCAGCAGCCTCAAGGAGCAGCCATGGCCCCAGACTTGTCGCACGGATGCAGAAAACTTAATGGAGGAGGCTGAGGTCAGAATGGGAAGAGTTTTTAAAAAATAAAAAGGGGAGCTAATATGTGAGGAACTGATTGCTGTGCATTATTTTCCACTGAACAGACCTTCTTTTAACTTCTAATCTCAAGCACATAGCAGCTGTTGTTTTCATATTAAAGATCAGTAAGTGTTTGTTGAATGAATGTATAAATTCTTCCCCAATCTCTGTTCCATGTAATAAGAACTAAGTTAGCATTCAATTATAAAGAGGACTTAGAAAGTCTGCAGACATTTTTGATCAAAAATCCTTTTAACTGTCTTCTTATGGTTCCCGGTAACACCTAAGCCCAAAAAACTCAGGTCATTTCTTGCCTGTACTGTTGCAATAGCCTCATAATAACTAGTCTTCTCCTTTATCAACCACAGTCGTCTCCGCTACACAAAGAAGAGAGGGTGCTAATCTAAAGTGTAAATTATTTTAAATCAGCACCCTTCTTGTAATTTCCCAGTGGTTTTTCCTGTCATAAGAATAACATGCAAAATTCCTTGCTTGCAAAGCCAGGTATGACCTTGCCCCGTCTACCTTACTGGCATCGTCTCTCATGACTCTTCCCCTTTTTCACTCTCCATCTTGCCTTTTCCCAACCATGCCTGCTAGTGAGCCTTAGCCCTTGTGTTTTCTCTGCCAGGGATGCATGACTGGCTTCTTTACTTTATTCTGATCTCTCTGCTCAAATGTCATCTTTTTAGAGATAGCTTTCCTAAACATCTTTAAAAGTACCCCATGAACTGTGAAACACCTTTCCTGCTTTTTCTTCATAGTACTTATCACTTCCTGATATTATGTTAGACAACAAGCAAATAAGTATATGCTGTTCCCACTAAAATATAAGCTCCATTTGAACAGGAGCCTTGATTATTCCTGAGTATACCCCTAGTGCCCAGAACCTAGTAGGGCTTTAATAAATATTAGATTCATTTGTTGAATGAAAAAATGAATGAATAATTGAATAAAGATATCCAGGTAGAGTGTCCGGGCAGCCCAACCTATGTGAAGCCAGTACAAGAATGTTTCTGTGACATAGTGAGCGCAATTATTCCTCTAAATGATGAACTTGCATGAGTACACAAGGATAAAATTAAGCACGTAGTTTAGATTTCTGCAAGTCTCTTCCTCAGATTGAGGAAAAACAAAGAAATTGAGCTTATTCTCACCTCATTCTTACATGAACTAAATCCAGAATAAAGCATCATTTAAATACCAAGCACATCCCTACCTCAGCCACTCATGGAATACTTTATCTGTAAAGGGACAGATAGATAGTAAATATTTTTGGCTTTGGGGACCGTACAGTCTGTGTCGCCACTACTCAGCACTTCTGTTGAGGTGTGGAATCAGCCCTAGACAACACATAAACAAATGGGTGTGGCTGTGTTCCAATAAAATGTTATTTATAAAACTAGGAAATCAGCAGATTTAGTCCACAGGCTGTAGTTTGCTGACCCTTGAATTAGAGAGTCTTCCCCAAATCCTCTTCCCAAGTTATTGAAAGAATGCCCATAGGACCATAGCTCCCTCACCTGGTTCTTTCAAGAAATGAGACAAATTCAGGGATGGTTGGAGGGTGGAACGCCTACCTGTGACATGACTATTGAGCCTGATTTCTTTCCTTTTATCTCTTTAAAGTGAGTCAGGTCTTAACTTTAAACTGTCTCTAAAGTGGCAATAACTGGGCCATGTCAAATACTCCCAATGACAAAATTTCTTCTTGGAGAATGAATGAGGGTTGCATACAGTGGTGTGTTGGTAAATGTTTAGCAACTGGCTCTCTTACTGATTTCTGTAGTGTAAATTCTCCCATCCTGGCCATCTGAAACTACCAATATGATGTCACTTGATGTAGAGTTGGGAAAAGACAGTCTCTTGAATAAGTATGAGCTGGTTCCAGCCCACTGCTGAACACATATTTAAAAGCTACAATAGCTCTGCTCACTGCCATTGCAGAGTCTGGAGGAAAGAGTAGCAGGCTTGATAGGAGCAATTGGCTGAATAATCTTTTGCAATGCCAATAAATAATGAAAAAGTATATAAATATAACGCTGTATAGAAAGCTCTCAAGTTCAGTGTTTGGTTAAAATACATACTCAGTAAATGGTAGCTATTATTGTCTTAGTTTAAGTTATTGCAAGCATTAAAATTAAATGTTTAGCTACAGACTCAATCCAGTTTTAATGTCATTGTGTTAATAAGGCCTCTTAACATTGAAGCAACAAAGATAAAGGAATTTTTTTGTAGCTATTCACGATTCTGCCCAGCTCAAGGTAAAGATAGCAACAAAATACTCCAACCAGAATAAAAGAAGTCTCCACTTGACTGCAGCTCTTCGCAACACTTCACAAGGGTCAAGCTGTGTTAGTTTCCTTCAGTGTTAAACAGAATGCATCTTAAGAGAATAGTTCAATTATTGATTGAATAGTAGGTTCAAGTTGACTATAATCCCAACTAATACACTCTAAGACCAGGGATGGGCCGACAGAAGGGTAAAGTTTTCCACTCAAGTCTCGAGGCAAACACTCAGATAGAAGTAGTAATATGGATGCCTTCTATTATTTGGTCACATACAATACTGTTCCAGCCCAGTTGTGCAAAACTGGTGAAATAAAAGGCAAAAATATTGACAACAGTTTGGGTAACTGGATTCCAAAATGGGCATCCATCTAAGAG\n",
      "0\n",
      "[exon\tENSE00001494061\t5019522\t5019548, exon\tENSE00001715598\t5020114\t5020398, exon\tENSE00001494058\t5130591\t5130687, exon\tENSE00001625593\t5180056\t5180115, exon\tENSE00001789031\t5436490\t5436538, exon\tENSE00001769076\t5486379\t5486439, exon\tENSE00001541870\t5543340\t5543454, exon\tENSE00003717383\t5555994\t5558775]\n",
      "[26, 310, 406, 465, 513, 573, 687, 3468]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[(0, 5019522, 26), (26, 5019548, 850), (876, 5020398, 2592)]"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import random\n",
    "isoform = random.sample(known_isoforms, 1)[0]\n",
    "iso_str = ''.join(map(lambda x: genome[x.start:x.end], isoform.exons))\n",
    "iso_str = list(iso_str)\n",
    "iso_str[7] = random.choice(ALPHABET)\n",
    "iso_str[25] = random.choice(ALPHABET)\n",
    "iso_str[56] = random.choice(ALPHABET)\n",
    "iso_str[89] = random.choice(ALPHABET)\n",
    "iso_str[101] = random.choice(ALPHABET)\n",
    "iso_str[120] = random.choice(ALPHABET)\n",
    "iso_str[140] = random.choice(ALPHABET)\n",
    "iso_str[167] = random.choice(ALPHABET)\n",
    "iso_str=''.join(iso_str)\n",
    "print(iso_str)\n",
    "#outer_align(iso_str)\n",
    "a.align(iso_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GCCACCACTGTGGTCTAAGATCTATCACCTGCAGAAGTGACCTGTGGCAA'"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome[8542892:8542892+50]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "s = 'ATGATCGTCTATCGATCTGACGGTATGCAGTCATCGACTATGATTGCGTATAGCAGTCAGCTGA$'\n",
    "sa = get_suffix_array(s)\n",
    "L = get_bwt(s,sa)\n",
    "M = get_M(get_F(L))\n",
    "occ = get_occ(L)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((13, 16), 0)"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bowtie('ATG', M, occ)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'A'"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "random.choice(ALPHABET)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(0, 8542892, 50)]"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a.align(reads[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATTACTCTTGGGAATGAAATCCTATCTATATAAGCTGTGGTTTGAAATCC'"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reads[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TTGCTTTTCCAGTCTCTTTGGTATCTGCCGTGAAAGCAAGATGACAAAGGGAACGCCCTTGTTTGTAGTGTTGCAATAACACTCGCTCTTGGGCCACCACTGTGGTCTAAGATCTATCACCTGCAGAAGTGACCTGTGGCAAGTGTGGCT'"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome[8542892:8542892+50]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "210\n",
      "[exon\tENSE00001788379\t6387890\t6387997, exon\tENSE00001712270\t6390677\t6390843, exon\tENSE00001679869\t6393083\t6393200, exon\tENSE00001713465\t6397865\t6398185]\n",
      "[107, 273, 390, 710]\n",
      "2\n",
      "3\n",
      "4\n",
      "23\n",
      "[exon\tENSE00001802701\t8250613\t8250877, exon\tENSE00001729938\t8252369\t8252739]\n",
      "[264, 634]\n",
      "5\n",
      "15\n",
      "[exon\tENSE00003716433\t2673189\t2673274]\n",
      "[85]\n",
      "6\n",
      "56\n",
      "[exon\tENSE00001807840\t1996583\t1996689]\n",
      "[106]\n",
      "7\n",
      "165\n",
      "[exon\tENSE00001646882\t2775696\t2776627]\n",
      "[931]\n",
      "8\n",
      "9\n",
      "320\n",
      "[exon\tENSE00001294445\t2725086\t2726266]\n",
      "[1180]\n",
      "10\n",
      "230\n",
      "[exon\tENSE00003756290\t2254152\t2254221, exon\tENSE00003755158\t2263408\t2263944]\n",
      "[69, 605]\n",
      "11\n",
      "314\n",
      "[exon\tENSE00001643453\t2718932\t2719213, exon\tENSE00001800469\t2720565\t2720740]\n",
      "[281, 456]\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-159-d281999415d5>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mcounter\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mread\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mreads\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m     \u001b[0malignment\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0ma\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0malign\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m     \u001b[0mevaluation\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mevaluate_alignment\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgenome\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mread\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0malignment\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0munknown_isoforms\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgenome_isoform_offsets\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0mevaluations\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mevaluation\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-154-2913ab65516c>\u001b[0m in \u001b[0;36malign\u001b[0;34m(self, read_sequence)\u001b[0m\n\u001b[1;32m    412\u001b[0m             \u001b[0mcurr_pt\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mexons\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mexon_num\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstart\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mcurr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    413\u001b[0m             \u001b[0;32mwhile\u001b[0m \u001b[0mseqlength\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 414\u001b[0;31m                 \u001b[0moffset\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mexons\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mexon_num\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mend\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mcurr_pt\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    415\u001b[0m                 \u001b[0mdiff\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moffset\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mseqlength\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    416\u001b[0m                 \u001b[0malignment\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mread_pt\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcurr_pt\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdiff\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "evaluations = []\n",
    "counter = 0\n",
    "for read in reads:\n",
    "    alignment = a.align(read)\n",
    "    evaluation = evaluate_alignment(genome, read, alignment, unknown_isoforms, genome_isoform_offsets)\n",
    "    evaluations.append(evaluation)\n",
    "    counter += 1\n",
    "    print(counter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({('hidden_gene', -50): 1,\n",
       "         ('hidden_gene', -48): 1,\n",
       "         ('hidden_gene', -46): 4,\n",
       "         ('hidden_gene', -44): 2,\n",
       "         ('hidden_gene', -42): 3,\n",
       "         ('hidden_gene', -40): 1,\n",
       "         ('unaligned', 0): 1563})"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from collections import Counter\n",
    "Counter(evaluations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "723\n"
     ]
    }
   ],
   "source": [
    "counter = 0\n",
    "def suffix_replacer(hits):\n",
    "    (start, end), length = hits\n",
    "    genome_seqs = []\n",
    "    for i in range(start,end):\n",
    "        index = a.genome_sa[i]\n",
    "        genome_seqs.append((index,index+length))\n",
    "    return genome_seqs\n",
    "def genome_align():\n",
    "    min_mis = 7\n",
    "    best_hits = []\n",
    "    hits = bowtie(read, a.genome_M, a.genome_occ)\n",
    "    curr_mis = hits[1]\n",
    "    if curr_mis < min_mis:\n",
    "        min_mis = curr_mis\n",
    "        best_hits = [(0, suffix_replacer(hits)[0][0], 50)]\n",
    "    for offset in range(10, 41):\n",
    "        seed1, seed2 = read[:offset], read[offset:]\n",
    "        hits1 = bowtie(seed1, a.genome_M, a.genome_occ)\n",
    "        hits2 = bowtie(seed2, a.genome_M, a.genome_occ)\n",
    "        curr_mis = hits1[1] + hits2[1]\n",
    "        if curr_mis < min_mis:\n",
    "            hits1 = suffix_replacer(hits1)\n",
    "            hits2 = suffix_replacer(hits2)\n",
    "            for seq1 in hits1:\n",
    "                for seq2 in hits2:\n",
    "                    if MIN_INTRON_SIZE < seq2[0] - seq1[1] < MAX_INTRON_SIZE:\n",
    "                        min_mis = curr_mis\n",
    "                        seq1len = len(seq1)\n",
    "                        seq2len = len(seq2)\n",
    "                        best_hits = [(0, seq1[0], seq1len), (seq1len, seq2[0], seq2len)]\n",
    "                        break\n",
    "    for offset1 in range(10,31):\n",
    "        for offset2 in range(10+offset1, 41):\n",
    "            seed1, seed2,seed3 = read[:offset1], read[offset1:offset2], read[offset2:]\n",
    "            hits1 = bowtie(seed1, a.genome_M, a.genome_occ)\n",
    "            hits2 = bowtie(seed2, a.genome_M, a.genome_occ)\n",
    "            hits3 = bowtie(seed3, a.genome_M, a.genome_occ)\n",
    "            curr_mis = hits1[1] + hits2[1] + hits3[1]\n",
    "            if curr_mis < min_mis:\n",
    "                hits1 = suffix_replacer(hits1)\n",
    "                hits2 = suffix_replacer(hits2)\n",
    "                hits3 = suffix_replacer(hits3)\n",
    "                for seq1 in hits1:\n",
    "                    for seq2 in hits2:\n",
    "                        for seq3 in hits3:\n",
    "                            if MIN_INTRON_SIZE < seq2[0] - seq1[1] < MAX_INTRON_SIZE and MIN_INTRON_SIZE < seq3[0] - seq2[1] < MAX_INTRON_SIZE:\n",
    "                                min_mis = curr_mis\n",
    "                                seq1len = len(seq1)\n",
    "                                seq2len = len(seq2)\n",
    "                                seq3len = len(seq3)\n",
    "                                best_hits = [(0, seq1[0], seq1len), (seq1len, seq2[0], seq2len), (seq1len+seq2len, seq3[0], seq3len)]\n",
    "                                break\n",
    "    return best_hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "counter = 0\n",
    "def suffix_replacer(hits):\n",
    "    (start, end), length = hits\n",
    "    genome_seqs = []\n",
    "    for i in range(start,end):\n",
    "        index = a.genome_sa[i]\n",
    "        genome_seqs.append((index,index+length))\n",
    "    return genome_seqs\n",
    "iteration = 0\n",
    "for read in reads:\n",
    "    min_mis = 7\n",
    "    best_hits = []\n",
    "    hits = bowtie(read, a.genome_M, a.genome_occ)\n",
    "    curr_mis = hits[1]\n",
    "    if curr_mis < min_mis:\n",
    "        min_mis = curr_mis\n",
    "        best_hits = suffix_replacer(hits)\n",
    "    for offset in range(8, 43):\n",
    "        seed1, seed2 = read[:offset], read[offset:]\n",
    "        hits1 = bowtie(seed1, a.genome_M, a.genome_occ)\n",
    "        hits2 = bowtie(seed2, a.genome_M, a.genome_occ)\n",
    "        curr_mis = hits1[1] + hits2[1]\n",
    "        if curr_mis < min_mis:\n",
    "            hits1 = suffix_replacer(hits1)\n",
    "            hits2 = suffix_replacer(hits2)\n",
    "            for seq1 in hits1:\n",
    "                for seq2 in hits2:\n",
    "                    if MIN_INTRON_SIZE < seq2[0] - seq1[1] < MAX_INTRON_SIZE:\n",
    "                        min_mis = curr_mis\n",
    "                        best_hits = hits\n",
    "    for offset1 in range(8,35):\n",
    "        for offset2 in range(8+offset1, 43):\n",
    "            seed1, seed2,seed3 = read[:offset1], read[offset1:offset2], read[offset2:]\n",
    "            hits1 = bowtie(seed1, a.genome_M, a.genome_occ)\n",
    "            hits2 = bowtie(seed2, a.genome_M, a.genome_occ)\n",
    "            hits3 = bowtie(seed3, a.genome_M, a.genome_occ)\n",
    "            curr_mis = hits1[1] + hits2[1] + hits3[1]\n",
    "            if curr_mis < min_mis:\n",
    "                hits1 = suffix_replacer(hits1)\n",
    "                hits2 = suffix_replacer(hits2)\n",
    "                hits3 = suffix_replacer(hits3)\n",
    "                for seq1 in hits1:\n",
    "                    for seq2 in hits2:\n",
    "                        for seq3 in hits3:\n",
    "                            if MIN_INTRON_SIZE < seq2[0] - seq1[1] < MAX_INTRON_SIZE and MIN_INTRON_SIZE < seq3[0] - seq2[1] < MAX_INTRON_SIZE:\n",
    "                                min_mis = curr_mis\n",
    "                                best_hits = hits\n",
    "    iteration += 1\n",
    "    if best_hits:\n",
    "        counter += 1\n",
    "    print('iteration: ', iteration, ', counter: ', counter, ', time taken: ', time.time()-start)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "counter = 0\n",
    "def suffix_replacer(hits):\n",
    "    (start, end), length = hits\n",
    "    genome_seqs = []\n",
    "    for i in range(start,end):\n",
    "        index = a.genome_sa[i]\n",
    "        genome_seqs.append((index,index+length))\n",
    "    return genome_seqs\n",
    "iteration = 0\n",
    "for read in reads:\n",
    "    start = time.time()\n",
    "    min_mis = 7\n",
    "    best_hits = []\n",
    "    before_single = time.time()\n",
    "    hits = bowtie(read, a.genome_M, a.genome_occ)\n",
    "    curr_mis = hits[1]\n",
    "    if curr_mis < min_mis:\n",
    "        min_mis = curr_mis\n",
    "        best_hits = suffix_replacer(hits)\n",
    "    print('single: ', time.time()-before_single)\n",
    "    before_double = time.time()\n",
    "    for offset in range(5, 46):\n",
    "        seed1, seed2 = read[:offset], read[offset:]\n",
    "        hits1 = bowtie(seed1, a.genome_M, a.genome_occ)\n",
    "        hits2 = bowtie(seed2, a.genome_M, a.genome_occ)\n",
    "        curr_mis = hits1[1] + hits2[1]\n",
    "        if curr_mis < min_mis:\n",
    "            hits1 = suffix_replacer(hits1)\n",
    "            hits2 = suffix_replacer(hits2)\n",
    "            for seq1 in hits1:\n",
    "                for seq2 in hits2:\n",
    "                    if MIN_INTRON_SIZE < seq2[0] - seq1[1] < MAX_INTRON_SIZE:\n",
    "                        min_mis = curr_mis\n",
    "                        best_hits = [hits1, hits2]\n",
    "                        break\n",
    "    print('double: ', time.time()-before_double)\n",
    "    before_triple = time.time()\n",
    "    for offset1 in range(5,41):\n",
    "        for offset2 in range(5+offset1, 46):\n",
    "            seed1, seed2,seed3 = read[:offset1], read[offset1:offset2], read[offset2:]\n",
    "            hits1 = bowtie(seed1, a.genome_M, a.genome_occ)\n",
    "            hits2 = bowtie(seed2, a.genome_M, a.genome_occ)\n",
    "            hits3 = bowtie(seed3, a.genome_M, a.genome_occ)\n",
    "            curr_mis = hits1[1] + hits2[1] + hits3[1]\n",
    "            if curr_mis < min_mis:\n",
    "                hits1 = suffix_replacer(hits1)\n",
    "                hits2 = suffix_replacer(hits2)\n",
    "                hits3 = suffix_replacer(hits3)\n",
    "                print('suffix replacing: ', time.time()-before_triple)\n",
    "                for seq1 in hits1:\n",
    "                    for seq2 in hits2:\n",
    "                        for seq3 in hits3:\n",
    "                            if MIN_INTRON_SIZE < seq2[0] - seq1[1] < MAX_INTRON_SIZE and MIN_INTRON_SIZE < seq3[0] - seq2[1] < MAX_INTRON_SIZE:\n",
    "                                min_mis = curr_mis\n",
    "                                best_hits = [hits1, hits2, hits3]\n",
    "                                break\n",
    "    print('triple: ', time.time()-before_triple)\n",
    "    iteration += 1\n",
    "    if best_hits:\n",
    "        counter += 1\n",
    "    print('iteration: ', iteration, ', counter: ', counter, ', time taken: ', time.time()-start)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
